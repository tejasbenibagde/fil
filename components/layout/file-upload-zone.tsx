"use client";

import {
  useState,
  useRef,
  type ChangeEvent,
  type DragEvent,
  type MouseEvent,
} from "react";
import { Button } from "@/components/ui/button";
import { Upload } from "lucide-react";

interface FileUploadZoneProps {
  onFilesSelected: (files: File[]) => void;
  accept?: string;
  maxFiles?: number;
  buttonText?: string;
  description?: string;
  className?: string;
  disabled?: boolean;
}

export function FileUploadZone({
  onFilesSelected,
  accept = "*/*",
  maxFiles = 1,
  buttonText = "Browse Files",
  description = "Drag and drop your file here, or click to browse",
  className = "",
  disabled = false,
}: FileUploadZoneProps) {
  const [isDragging, setIsDragging] = useState(false);
  const fileInputRef = useRef<HTMLInputElement>(null);
  const containerRef = useRef<HTMLDivElement>(null);

  const handleDragEnter = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    if (!disabled) setIsDragging(true);
  };

  const handleDragOver = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    if (!disabled && !isDragging) setIsDragging(true);
  };

  const handleDragLeave = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();

    // Only set isDragging to false if we're leaving the container (not entering a child element)
    if (!e.currentTarget.contains(e.relatedTarget as Node)) {
      setIsDragging(false);
    }
  };

  const handleDrop = (e: DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(false);

    if (disabled) return;

    const files = Array.from(e.dataTransfer.files);
    processFiles(files);
  };

  const handleFileInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (disabled || !e.target.files?.length) return;

    const files = Array.from(e.target.files);
    processFiles(files);
  };

  const processFiles = (files: File[]) => {
    // Filter files based on accept attribute
    let filteredFiles = files;

    if (accept !== "*/*") {
      const acceptedTypes = accept.split(",").map((type) => type.trim());
      filteredFiles = files.filter((file) => {
        return acceptedTypes.some((type) => {
          // Handle mime types (e.g., "image/*")
          if (type.endsWith("/*")) {
            const category = type.split("/")[0];
            return file.type.startsWith(`${category}/`);
          }
          // Handle specific extensions (e.g., ".pdf")
          if (type.startsWith(".")) {
            return file.name.toLowerCase().endsWith(type.toLowerCase());
          }
          // Handle exact mime types
          return file.type === type;
        });
      });
    }

    // Limit number of files
    if (maxFiles > 0 && filteredFiles.length > maxFiles) {
      filteredFiles = filteredFiles.slice(0, maxFiles);
    }

    if (filteredFiles.length > 0) {
      onFilesSelected(filteredFiles);
    }
  };

  const handleContainerClick = (e: MouseEvent<HTMLDivElement>) => {
    // Only trigger if the click is directly on the container or specific child elements
    // but not on the button or its children
    if (disabled) return;

    // Check if the click target is the button or a child of the button
    const buttonElement = containerRef.current?.querySelector("button");
    if (
      buttonElement &&
      (buttonElement === e.target || buttonElement.contains(e.target as Node))
    ) {
      return; // Don't do anything if clicking on the button
    }

    // Otherwise, trigger the file input
    if (fileInputRef.current) {
      fileInputRef.current.click();
    }
  };

  const triggerFileInput = () => {
    if (!disabled && fileInputRef.current) {
      fileInputRef.current.click();
    }
  };

  return (
    <div
      ref={containerRef}
      className={`border-2 ${
        isDragging ? "border-primary" : "border-dashed"
      } rounded-lg p-8 text-center transition-colors ${className} ${
        disabled ? "opacity-60 cursor-not-allowed" : "cursor-pointer"
      }`}
      onDragEnter={handleDragEnter}
      onDragOver={handleDragOver}
      onDragLeave={handleDragLeave}
      onDrop={handleDrop}
      onClick={handleContainerClick}
      tabIndex={0}
      role="button"
      aria-disabled={disabled}
    >
      <div className="flex flex-col items-center">
        <Upload
          className={`h-10 w-10 mb-4 ${
            isDragging ? "text-primary" : "text-muted-foreground"
          }`}
        />
        <h3 className="text-lg font-medium mb-2">
          {isDragging ? "Drop your file here" : "Upload your file"}
        </h3>
        <p className="text-sm text-muted-foreground mb-4">{description}</p>

        {/* Completely separate button without label wrapping */}
        <Button
          aria-label="upload button"
          disabled={disabled}
          onClick={(e) => {
            e.stopPropagation();
            triggerFileInput();
          }}
        >
          {buttonText}
        </Button>

        {/* Hidden file input without label */}
        <input
          data-testid="file-input"
          type="file"
          className="sr-only"
          accept={accept}
          onChange={handleFileInputChange}
          ref={fileInputRef}
          multiple={maxFiles > 1}
          disabled={disabled}
        />
      </div>
    </div>
  );
}
