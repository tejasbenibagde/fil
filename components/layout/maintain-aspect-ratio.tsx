"use client"

import { Button } from "@/components/ui/button"
import { Lock, Unlock } from "lucide-react"

interface MaintainAspectRatioProps {
  maintainAspectRatio: boolean
  toggleAspectRatio: () => void
  originalWidth: number | null
  originalHeight: number | null
  resetDimensions: () => void
  disabled?: boolean
}

export function MaintainAspectRatio({
  maintainAspectRatio,
  toggleAspectRatio,
  originalWidth,
  originalHeight,
  resetDimensions,
  disabled = false,
}: MaintainAspectRatioProps) {
  return (
    <div className="flex items-center justify-between mb-2">
      <div className="flex items-center">
        <Button
          variant="ghost"
          size="icon"
          onClick={toggleAspectRatio}
          className="mr-2"
          title={maintainAspectRatio ? "Unlock aspect ratio" : "Lock aspect ratio"}
          disabled={disabled}
          aria-label="toggle aspect ratio"
        >
          {maintainAspectRatio ? <Lock className="h-4 w-4 text-primary" /> : <Unlock className="h-4 w-4" />}
        </Button>
        <span className="text-sm font-medium">{maintainAspectRatio ? "Maintain aspect ratio" : "Free resize"}</span>
      </div>
      {originalWidth && originalHeight && (
        <Button variant="outline" size="sm" onClick={resetDimensions} disabled={disabled}>
          Reset to Original
        </Button>
      )}
    </div>
  )
}
