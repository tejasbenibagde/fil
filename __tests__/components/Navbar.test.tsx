import { Navbar } from "@/components/layout/navbar";
import { render } from "@testing-library/react";
import { usePathname } from "next/navigation";
jest.mock("next/navigation", () => {
  return {
    ...jest.requireActual("next/navigation"),
    usePathname: jest.fn(),
  };
});

describe("Navbar", () => {
  beforeEach(() => {
    (usePathname as jest.Mock).mockReturnValue("/");
  });
  it("should render navbar and match snapshot", () => {
    const navbar = render(<Navbar />);
    expect(navbar).toMatchSnapshot();
  });
});
