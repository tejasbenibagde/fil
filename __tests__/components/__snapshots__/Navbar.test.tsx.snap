// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render navbar and match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="sticky top-0 z-40 w-full border-b bg-background"
      >
        <div
          class="container flex h-16 items-center space-x-4 sm:justify-between sm:space-x-0"
        >
          <div
            class="flex gap-6 md:gap-10"
          >
            <a
              class="flex items-center space-x-2"
              href="/"
            >
              <img
                alt="Fil Logo"
                class="mix-blend-difference"
                data-nimg="1"
                decoding="async"
                height="48"
                src="/fil-white.svg"
                style="color: transparent;"
                width="48"
              />
            </a>
            <nav
              class="hidden gap-6 md:flex"
            >
              <a
                class="flex items-center text-sm font-medium text-muted-foreground transition-colors hover:text-primary"
                href="/tools"
              >
                All Tools
              </a>
              <a
                class="flex items-center text-sm font-medium text-muted-foreground transition-colors hover:text-primary"
                href="/about"
              >
                About
              </a>
            </nav>
          </div>
          <div
            class="flex flex-1 items-center justify-end space-x-4"
          >
            <nav
              class="flex items-center space-x-2"
            >
              <button
                class="inline-flex items-center cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9"
                data-slot="button"
              >
                <svg
                  class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="4"
                  />
                  <path
                    d="M12 2v2"
                  />
                  <path
                    d="M12 20v2"
                  />
                  <path
                    d="m4.93 4.93 1.41 1.41"
                  />
                  <path
                    d="m17.66 17.66 1.41 1.41"
                  />
                  <path
                    d="M2 12h2"
                  />
                  <path
                    d="M20 12h2"
                  />
                  <path
                    d="m6.34 17.66-1.41 1.41"
                  />
                  <path
                    d="m19.07 4.93-1.41 1.41"
                  />
                </svg>
                <svg
                  class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
                  />
                </svg>
                <span
                  class="sr-only"
                >
                  Toggle theme
                </span>
              </button>
              <a
                href="https://github.com/tejasbenibagde/fil"
                rel="noreferrer"
                target="_blank"
              >
                <button
                  class="inline-flex items-center cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9"
                  data-slot="button"
                >
                  <svg
                    class="lucide lucide-github h-4 w-4"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
                    />
                    <path
                      d="M9 18c-4.51 2-5-2-7-2"
                    />
                  </svg>
                  <span
                    class="sr-only"
                  >
                    GitHub
                  </span>
                </button>
              </a>
              <button
                aria-controls="radix-«r0»"
                aria-expanded="false"
                aria-haspopup="dialog"
                class="inline-flex items-center cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 md:hidden"
                data-slot="sheet-trigger"
                data-state="closed"
                type="button"
              >
                <svg
                  class="lucide lucide-menu h-5 w-5"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="4"
                    x2="20"
                    y1="12"
                    y2="12"
                  />
                  <line
                    x1="4"
                    x2="20"
                    y1="6"
                    y2="6"
                  />
                  <line
                    x1="4"
                    x2="20"
                    y1="18"
                    y2="18"
                  />
                </svg>
                <span
                  class="sr-only"
                >
                  Toggle menu
                </span>
              </button>
            </nav>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="sticky top-0 z-40 w-full border-b bg-background"
    >
      <div
        class="container flex h-16 items-center space-x-4 sm:justify-between sm:space-x-0"
      >
        <div
          class="flex gap-6 md:gap-10"
        >
          <a
            class="flex items-center space-x-2"
            href="/"
          >
            <img
              alt="Fil Logo"
              class="mix-blend-difference"
              data-nimg="1"
              decoding="async"
              height="48"
              src="/fil-white.svg"
              style="color: transparent;"
              width="48"
            />
          </a>
          <nav
            class="hidden gap-6 md:flex"
          >
            <a
              class="flex items-center text-sm font-medium text-muted-foreground transition-colors hover:text-primary"
              href="/tools"
            >
              All Tools
            </a>
            <a
              class="flex items-center text-sm font-medium text-muted-foreground transition-colors hover:text-primary"
              href="/about"
            >
              About
            </a>
          </nav>
        </div>
        <div
          class="flex flex-1 items-center justify-end space-x-4"
        >
          <nav
            class="flex items-center space-x-2"
          >
            <button
              class="inline-flex items-center cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9"
              data-slot="button"
            >
              <svg
                class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="4"
                />
                <path
                  d="M12 2v2"
                />
                <path
                  d="M12 20v2"
                />
                <path
                  d="m4.93 4.93 1.41 1.41"
                />
                <path
                  d="m17.66 17.66 1.41 1.41"
                />
                <path
                  d="M2 12h2"
                />
                <path
                  d="M20 12h2"
                />
                <path
                  d="m6.34 17.66-1.41 1.41"
                />
                <path
                  d="m19.07 4.93-1.41 1.41"
                />
              </svg>
              <svg
                class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
                />
              </svg>
              <span
                class="sr-only"
              >
                Toggle theme
              </span>
            </button>
            <a
              href="https://github.com/tejasbenibagde/fil"
              rel="noreferrer"
              target="_blank"
            >
              <button
                class="inline-flex items-center cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9"
                data-slot="button"
              >
                <svg
                  class="lucide lucide-github h-4 w-4"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
                  />
                  <path
                    d="M9 18c-4.51 2-5-2-7-2"
                  />
                </svg>
                <span
                  class="sr-only"
                >
                  GitHub
                </span>
              </button>
            </a>
            <button
              aria-controls="radix-«r0»"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="inline-flex items-center cursor-pointer justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 md:hidden"
              data-slot="sheet-trigger"
              data-state="closed"
              type="button"
            >
              <svg
                class="lucide lucide-menu h-5 w-5"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="4"
                  x2="20"
                  y1="12"
                  y2="12"
                />
                <line
                  x1="4"
                  x2="20"
                  y1="6"
                  y2="6"
                />
                <line
                  x1="4"
                  x2="20"
                  y1="18"
                  y2="18"
                />
              </svg>
              <span
                class="sr-only"
              >
                Toggle menu
              </span>
            </button>
          </nav>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
